src/com/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -883,6 +884,108 @@
+        INNER JOIN fact_paths fp on fv.path_id = fp.id"
+
+   "DROP TABLE facts"
+
+   ;; Remove all the orphaned duplicates (all but the row in each set
+   ;; with min-id).
+   (if (scf-utils/postgres?)
+     "DELETE FROM fact_paths t1 USING fact_paths t2
+        WHERE t1.path = t2.path AND t1.id > t2.id"
+     "DELETE FROM fact_paths t1
+        WHERE t1.id <> (SELECT MIN(t2.id) FROM fact_paths t2
+                          WHERE t1.path = t2.path)")
+   (if (scf-utils/postgres?)
+     "DELETE FROM fact_values t1 USING fact_values t2
+        WHERE t1.value_hash = t2.value_hash AND t1.id > t2.id"
+     "DELETE FROM fact_values t1
+        WHERE t1.id <> (SELECT MIN(t2.id) FROM fact_values t2
+                          WHERE t1.value_hash = t2.value_hash)")
+
+   (let [opt-deferrable (if (and (scf-utils/postgres?)
+                                 (scf-utils/db-version-newer-than? [9 0]))
+                          "DEFERRABLE" "")]
+     (format "ALTER TABLE fact_paths
+                ADD CONSTRAINT fact_paths_path_key UNIQUE (path)
+                  %s" opt-deferrable)
+     (format "ALTER TABLE fact_values
+                ADD CONSTRAINT fact_values_value_hash_key UNIQUE (value_hash)
+                  %s" opt-deferrable))
+
+   "CREATE TABLE facts_transform
+      (factset_id bigint NOT NULL,
+       fact_path_id bigint NOT NULL,
+       fact_value_id bigint NOT NULL)"
+
+   ;; Patch up facts refrences to refer to the min id path/value
+   ;; wherever there's more than one option.
+   "INSERT INTO facts_transform (factset_id, fact_path_id, fact_value_id)
+      SELECT f.factset_id, fp.id, fv.id
+        FROM facts_unique_transform f
+        INNER JOIN fact_paths fp on f.fact_path = fp.path
+        INNER JOIN fact_values fv on f.fact_value_hash = fv.value_hash;"
+
+   "DROP TABLE facts_unique_transform"
+
+   "ALTER TABLE facts_transform
+      ADD CONSTRAINT facts_factset_id_fact_path_id_fact_key
+        UNIQUE (factset_id, fact_path_id)"
+
+   "ALTER TABLE facts_transform
+      ADD CONSTRAINT factset_id_fk
+        FOREIGN KEY (factset_id) REFERENCES factsets(id)
+        ON UPDATE CASCADE
+        ON DELETE CASCADE"
+   ;; CHECK: Do we still want/need these to be deferrable, after the GC changes?
+   (let [opt-deferrable (if (scf-utils/postgres?) "DEFERRABLE" "")]
+     (format
+      "ALTER TABLE facts_transform
+         ADD CONSTRAINT fact_path_id_fk
+           FOREIGN KEY (fact_path_id) REFERENCES fact_paths(id)
+           %s" opt-deferrable)
+     (format
+      "ALTER TABLE facts_transform
+         ADD CONSTRAINT fact_value_id_fk
+           FOREIGN KEY (fact_value_id) REFERENCES fact_values(id)
+           %s" opt-deferrable))
+
+   "ALTER TABLE facts_transform RENAME TO facts"
+   "CREATE INDEX facts_fact_path_id_idx ON facts(fact_path_id)"
+   "CREATE INDEX facts_fact_value_id_idx ON facts(fact_value_id)"
+
+   (if (scf-utils/postgres?) "ANALYZE facts" "SELECT 1")
+
+   ;; These are for the more pedantic HSQLDB.
```

Comment:
@wkalt I found that HSQLDB required that we do that explicitly where PostgreSQL doesn't.